In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [3]:
wvs_w6 = pd.read_csv('../data/Evaluating_Happiness/wvs_w6.csv', low_memory=False)

In [4]:
wvs_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V44_ES  V45  V46  V47  V48  V49  V50  V51  \
0  2.0  1.0  2.0  2.0  1.0  1.0     NaN  1.0  NaN  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0     NaN  1.0  NaN  1.0  1.0  2.0  2.0  1.0   
2  2.0  2.0  1.0  1.0  1.0  2.0     NaN  1.0  NaN  2.0  3.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0     NaN  2.0  NaN  1.0  1.0  1.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0     NaN  1.0  NaN  1.0  1.0  1.0  2.0  1.0   

   V52  V53  V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  \
0  1.0  1.0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0   
1  2.0  3.0  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0   
2  1.0  1.0  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0   
3  2.0  2.0  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0   
4  3.0  2.0  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0   

   V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0   
1  3.0  1.0  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0   
2  4.0  1.0  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0   
3  3.0  1.0  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0   
4  2.0  1.0  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0   

   V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  \
0  3.0  2.0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  2.0  2.0  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2  2.0  2.0  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  NaN  NaN  NaN  NaN   
3  3.0  1.0  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4  3.0  1.0  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  \
0  NaN  5.0  8.0  7.0  6.0  8.0   7.0   5.0   2.0   2.0   2.0   2.0   2.0   
1  NaN  5.0  7.0  5.0  5.0  4.0   4.0   6.0   1.0   3.0   3.0   3.0   3.0   
2  NaN  6.0  7.0  7.0  7.0  5.0   7.0   5.0   1.0   3.0   3.0   4.0   4.0   
3  NaN  6.0  9.0  5.0  6.0  4.0   6.0   8.0   1.0   3.0   3.0   2.0   2.0   
4  NaN  5.0  8.0  4.0  7.0  4.0   6.0   6.0   2.0   2.0   3.0   4.0   2.0   

   V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  \
0   2.0   1.0   3.0   3.0   2.0   3.0   2.0   2.0   4.0   3.0   3.0   2.0   
1   3.0   1.0   1.0   1.0   2.0

In [5]:
wvs_w6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Columns: 442 entries, V1 to WEIGHT4B
dtypes: float64(428), int64(12), object(2)
memory usage: 302.0+ MB


In [ ]:
# possibly change dtypes to reduce memory usage
# what's the most efficient way to do this for 400+ variables?

In [6]:
wvs_w6.shape

(89565, 442)

In [7]:
wvs_w6.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Data columns (total 442 columns):
 #   Column           Dtype  
---  ------           -----  
 0   V1               int64  
 1   V2               int64  
 2   V2A              int64  
 3   cow              int64  
 4   C_COW_ALPHA      object 
 5   B_COUNTRY_ALPHA  object 
 6   V3               int64  
 7   V4               float64
 8   V5               float64
 9   V6               float64
 10  V7               float64
 11  V8               float64
 12  V9               float64
 13  V10              float64
 14  V11              float64
 15  V12              float64
 16  V13              float64
 17  V14              float64
 18  V15              float64
 19  V16              float64
 20  V17              float64
 21  V18              float64
 22  V19              float64
 23  V20              float64
 24  V21              float64
 25  V22              float64
 26  V23              float64
 27  V24            

Codebook available here: https://www.worldvaluessurvey.org/WVSDocumentationWV6.jsp (accessed on 5/25/2021)

In [8]:
# how many columns with null values?
wvs_w6.isna().sum().count()

442

In [9]:
# another approach for finding the number of columns with null values
sum(map(any, wvs_w6.isnull()))

442

In [10]:
# how many null values in each column?
#pd.set_option('display.max_rows', None)
wvs_w6.isna().sum()

V1                     0
V2                     0
V2A                    0
cow                    0
C_COW_ALPHA            0
B_COUNTRY_ALPHA        0
V3                     0
V4                   341
V5                   545
V6                  1035
V7                  1620
V8                  1519
V9                  1348
V10                  758
V11                  326
V12                   14
V13                    6
V14                    7
V15                   17
V16                    9
V17                   13
V18                   12
V19                   18
V20                   15
V21                   19
V22                    3
V23                  586
V24                 2388
V25                  840
V26                  923
V27                  993
V28                 3211
V29                 3129
V30                 1118
V31                 2223
V32                 1055
V33                 1214
V34                 1197
V35                 9696
V36                 3979


In [11]:
# how many null values in all?
wvs_w6.isna().sum().sum()

10721335

In [12]:
# how many columns contain missing values?
print(np.count_nonzero(wvs_w6.isna().sum() != 0), 'columns contain missing values')

# how many columns do not?
print(np.count_nonzero(wvs_w6.isna().sum() == 0), 'columns DO NOT contain missing values')

414 columns contain missing values
28 columns DO NOT contain missing values


#### Observations so far:
This data set contains 28 columns without any missing information, and 414 columns (out of 442) that contain at least 1 missing value.

#### Next steps:
- figure out if these missing values matter for my analyses
- decide what to do about missing values

In [13]:
# list columns containing missing data
wvs_w6.columns[wvs_w6.isna().any()].tolist()

['V4',
 'V5',
 'V6',
 'V7',
 'V8',
 'V9',
 'V10',
 'V11',
 'V12',
 'V13',
 'V14',
 'V15',
 'V16',
 'V17',
 'V18',
 'V19',
 'V20',
 'V21',
 'V22',
 'V23',
 'V24',
 'V25',
 'V26',
 'V27',
 'V28',
 'V29',
 'V30',
 'V31',
 'V32',
 'V33',
 'V34',
 'V35',
 'V36',
 'V37',
 'V38',
 'V39',
 'V40',
 'V41',
 'V42',
 'V43',
 'V44',
 'V44_ES',
 'V45',
 'V46',
 'V47',
 'V48',
 'V49',
 'V50',
 'V51',
 'V52',
 'V53',
 'V54',
 'V55',
 'V56',
 'V56_NZ',
 'V57',
 'V58',
 'V59',
 'V60',
 'V61',
 'V62',
 'V63',
 'V64',
 'V65',
 'V66',
 'V67',
 'V68',
 'V69',
 'V70',
 'V71',
 'V72',
 'V73',
 'V74',
 'V74B',
 'V75',
 'V76',
 'V77',
 'V78',
 'V79',
 'V80',
 'V81',
 'V82',
 'V83',
 'V84',
 'V85',
 'V86',
 'V87',
 'V88',
 'V89',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V95',
 'V96',
 'V97',
 'V98',
 'V99',
 'V100',
 'V101',
 'V102',
 'V103',
 'V104',
 'V105',
 'V106',
 'V107',
 'V108',
 'V109',
 'V110',
 'V111',
 'V112',
 'V113',
 'V114',
 'V115',
 'V116',
 'V117',
 'V118',
 'V119',
 'V120',
 'V121',
 'V122',


### Options for handling missing data:  
Advice from https://heartbeat.fritz.ai/data-handling-scenarios-part-2-working-with-missing-values-in-a-dataset-34b758cfc9fa and https://analyticsindiamag.com/5-ways-handle-missing-values-machine-learning-datasets/  
  
**Drop rows and columns with missing data**  
1. pros: 
  - quick and easy approach 
2. cons: 
  - could mean losing important information about the data along with the missing values
  - when using your model in production, the model will not automatically know how to handle missing data
3. when this makes sense: 
  - rows containing NULL values (missing values) are around 5% (or less) of the total data
  - NULL values in columns are significantly more than the other values present; it wouldn’t make sense to keep these, as they hold little or no descriptive information about the data  
  
**Mean/Median (numerical) & Mode (categorical) imputation**  
1. pros: 
  - easy to do
  - can be integrated into production or for a future unknown dataset
2. cons: 
  - distorts the distribution of the dataset
  - distorts the variance and covariance of the dataset
  - for mode imputation, may lead to an over-representation of the most frequent label if the missing values are quite large
3. when this makes sense: 
  - mean imputation works best for normally distributed distributions
  - median is better for skewed distributions 
  - mode imputation for categorical data works best if the missing values are missing at random
  - best to use this method when the missing values are around 5% (or less) of the total data
  
**Systematic Random Sampling Imputation**  
1. pros: 
  - does not distort variance or distribution 
2. cons: 
  - when replacing missing values in the test set as well, the imputed values from the train set will need to be stored in memory
3. when this makes sense: 
  - can be applied to both numerical and categorical variables
  - used when the values are missing at random
  - when we want to be able to reproduce the same value every time the variable is used (by using a random state)
  
### Remaining Questions:
Is "stratified" mean/median imputation an option?

In [14]:
# Columns to drop
# drop a handful of item scales at the end of the data set since a) these come from items that appear earlier in the survey,
# and b) they don't necessarily reduce the missing values count my much
# also dropping associated weights
cleaning_w6 = wvs_w6.drop(['Y001_1',
 'Y001_2',
 'Y001_3',
 'Y001_4',
 'Y001_5',
 'Y001',
 'Y002',
 'Y003',
 'sacsecval',
 'secvalwgt',
 'resemaval',
 'weightb',
 'I_AUTHORITY',
 'I_NATIONALISM',
 'I_DEVOUT',
 'defiance',
 'I_RELIGIMP',
 'I_RELIGBEL',
 'I_RELIGPRAC',
 'disbelief',
 'I_NORM1',
 'I_NORM2',
 'I_NORM3',
 'relativism',
 'I_TRUSTARMY',
 'I_TRUSTPOLICE',
 'I_TRUSTCOURTS',
 'scepticism',
 'I_INDEP', 
 'I_IMAGIN',
 'I_NONOBED', 
 'autonomy', 
 'I_WOMJOB',
 'I_WOMPOL',
 'I_WOMEDU',
 'equality',
 'I_HOMOLIB',
 'I_ABORTLIB',
 'I_DIVORLIB',
 'choice',
 'I_VOICE1',
 'I_VOICE2',
 'I_VOI2_00',
 'voice',
 'WEIGHT1A', 
 'WEIGHT2A', 
 'WEIGHT3A', 
 'WEIGHT4A',
 'WEIGHT1B', 
 'WEIGHT2B', 
 'WEIGHT3B', 
 'WEIGHT4B'], axis=1)

In [ ]:
# dropping MENA module items
cleaning_w6 = cleaning_w6.drop(['MN_35A',
 'MN_163A',
 'MN_163B',
 'MN_163C',
 'MN_228L',
 'MN_228M',
 'MN_228N',
 'MN_228O',
 'MN_228P',
 'MN_228Q',
 'MN_228R',
 'MN_228S1',
 'MN_228S2',
 'MN_228S3',
 'MN_228S4',
 'MN_228S5',
 'MN_228S6',
 'MN_228S7',
 'MN_228S8',
 'MN_229A',
 'MN_229B',
 'MN_230A',
 'MN_233A',
 'MN_233B',
 'MN_234A',
 'MN_237A',
 'MN_237B1',
 'MN_237B2',
 'MN_237B3',
 'MN_237B4',
 'MN_237B5',
 'MN_237B6',
 'MN_237B7',
 'MN_237C1',
 'MN_237C2',
 'MN_237C3',
 'MN_237C4',
 'MN_237C5',
 'MN_237C6',
 'MN_249A1',
 'MN_249A2',
 'MN_249A3'], axis=1)

In [15]:
# dropping interviewer observations and other items with high proportion of missings
cleaning_w6 = cleaning_w6.drop(['V228',
 'V228_LOCAL',
 'V228_2',
 'V228A',
 'V228B',
 'V228C',
 'V228D',
 'V228E',
 'V228F',
 'V228G',
 'V228H',
 'V228I',
 'V228J',
 'V228K',
 'V230',
 'V231',
 'V232',
 'V233',
 'V234',
# unhelpful/redundant items or variables with > 10% missing responses
 'V236',
 'V237',
 'V239',
 'V241', # year of birth
# these X003 items are age groups; drop them since I'll use and scale the actual age values (from V242) anyway
 'X003R2', 
 'X003R',
# immigrant status items; unfortunately in most cases at least 10% or more observations are missing 
 'V243',
 'V243_AU',
 'V244',
 'V244_AU',
 'V245',
 'V246',
 'V247',
# keeping highest education level; fewer than 1000 observations missing; may be able to impute missing values
#'V248', 
 'V248_CS',
 'V249',
 'V250',
 'V251',
 'V252',
 'V253',
 'V253_CS',
 'V254',
 'V255',
 'N_REGION_ISO',
 'V256',
 'V256B',
 'V256C',
 'V257',
 'V261',
 'V265'], axis=1)

In [16]:
# ensure that columns are being dropped so far
cleaning_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V44_ES  V45  V46  V47  V48  V49  V50  V51  \
0  2.0  1.0  2.0  2.0  1.0  1.0     NaN  1.0  NaN  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0     NaN  1.0  NaN  1.0  1.0  2.0  2.0  1.0   
2  2.0  2.0  1.0  1.0  1.0  2.0     NaN  1.0  NaN  2.0  3.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0     NaN  2.0  NaN  1.0  1.0  1.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0     NaN  1.0  NaN  1.0  1.0  1.0  2.0  1.0   

   V52  V53  V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  \
0  1.0  1.0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0   
1  2.0  3.0  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0   
2  1.0  1.0  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0   
3  2.0  2.0  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0   
4  3.0  2.0  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0   

   V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0   
1  3.0  1.0  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0   
2  4.0  1.0  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0   
3  3.0  1.0  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0   
4  2.0  1.0  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0   

   V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  \
0  3.0  2.0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  2.0  2.0  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2  2.0  2.0  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  NaN  NaN  NaN  NaN   
3  3.0  1.0  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4  3.0  1.0  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  \
0  NaN  5.0  8.0  7.0  6.0  8.0   7.0   5.0   2.0   2.0   2.0   2.0   2.0   
1  NaN  5.0  7.0  5.0  5.0  4.0   4.0   6.0   1.0   3.0   3.0   3.0   3.0   
2  NaN  6.0  7.0  7.0  7.0  5.0   7.0   5.0   1.0   3.0   3.0   4.0   4.0   
3  NaN  6.0  9.0  5.0  6.0  4.0   6.0   8.0   1.0   3.0   3.0   2.0   2.0   
4  NaN  5.0  8.0  4.0  7.0  4.0   6.0   6.0   2.0   2.0   3.0   4.0   2.0   

   V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  \
0   2.0   1.0   3.0   3.0   2.0   3.0   2.0   2.0   4.0   3.0   3.0   2.0   
1   3.0   1.0   1.0   1.0   2.0

In [17]:
# original shape for comparison
wvs_w6.shape

(89565, 442)

In [18]:
# current shape so far
cleaning_w6.shape

(89565, 342)

In [19]:
# continuing to evaluate columns with missing values
cleaning_w6.isna().sum()

V1                     0
V2                     0
V2A                    0
cow                    0
C_COW_ALPHA            0
B_COUNTRY_ALPHA        0
V3                     0
V4                   341
V5                   545
V6                  1035
V7                  1620
V8                  1519
V9                  1348
V10                  758
V11                  326
V12                   14
V13                    6
V14                    7
V15                   17
V16                    9
V17                   13
V18                   12
V19                   18
V20                   15
V21                   19
V22                    3
V23                  586
V24                 2388
V25                  840
V26                  923
V27                  993
V28                 3211
V29                 3129
V30                 1118
V31                 2223
V32                 1055
V33                 1214
V34                 1197
V35                 9696
V36                 3979


In [20]:
# fast forward this process...drop columns with 50% or more missing observations
high_missings_list = cleaning_w6.isna().sum()
type(high_missings_list)

pandas.core.series.Series

In [21]:
high_missings_list = pd.DataFrame(high_missings_list)
type(high_missings_list)

pandas.core.frame.DataFrame

In [22]:
high_missings_list = high_missings_list.reset_index()

In [23]:
high_missings_list = high_missings_list.rename(columns = {'index':'variable', 0:'NA_count'})

In [24]:
type(high_missings_list)

pandas.core.frame.DataFrame

In [25]:
high_missings_list.head()

variable  NA_count
0           V1         0
1           V2         0
2          V2A         0
3          cow         0
4  C_COW_ALPHA         0

In [26]:
# only select rows with 49000 or more missing observations
high_missings_list = high_missings_list[high_missings_list['NA_count'] > 44782]

In [27]:
# return these variable names as a list
high_missings_list = high_missings_list.variable.tolist()
high_missings_list

['V44_ES',
 'V56_NZ',
 'V74B',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V125_00',
 'V125_01',
 'V125_02',
 'V125_03',
 'V125_04',
 'V125_05',
 'V125_06',
 'V125_07',
 'V125_08',
 'V125_09',
 'V125_10',
 'V125_11',
 'V125_12',
 'V125_13',
 'V125_14',
 'V125_15',
 'V160A',
 'V160B',
 'V160C',
 'V160D',
 'V160E',
 'V160F',
 'V160G',
 'V160H',
 'V160I',
 'V160J',
 'V207A',
 'V215_01',
 'V215_02',
 'V215_03',
 'V215_04',
 'V215_05',
 'V215_06',
 'V215_07',
 'V215_08',
 'V215_10',
 'V215_11',
 'V215_12',
 'V215_13',
 'V215_14',
 'V215_15',
 'V215_16',
 'V215_17',
 'V215_18',
 'V218_ESMA',
 'V217_ESMA',
 'V219_ESMA',
 'V220_ESMA',
 'V221_ESMA',
 'V222_ESMA',
 'V223_ESMA',
 'V224_ESMA',
 'MN_35A',
 'MN_163A',
 'MN_163B',
 'MN_163C',
 'MN_228L',
 'MN_228M',
 'MN_228N',
 'MN_228O',
 'MN_228P',
 'MN_228Q',
 'MN_228R',
 'MN_228S1',
 'MN_228S2',
 'MN_228S3',
 'MN_228S4',
 'MN_228S5',
 'MN_228S6',
 'MN_228S7',
 'MN_228S8',
 'MN_229A',
 'MN_229B',
 'MN_230A',
 'MN_233A',
 'MN_233B',
 'MN_234A',
 'MN

In [28]:
len(high_missings_list)

102

#### Re-insert variables that shouldn't be dropped
- V74B
- V125_00 - V125_15; use these to create a "confidence in political/economic union" variable (V125_C)
- V215_01 - V215_18; use these to create a "sense of belonging to political/economic union" variable (V215_B)

In [29]:
# create a for loop to re-insert variables into the list that shouldn't be dropped (at this point)
# original list: high_missings_list
# new list
vars_to_keep = ['V74B', 'V125_00', 'V125_01', 'V125_02', 'V125_03', 'V125_04', 'V125_05', 'V125_06', 'V125_07', 'V125_08',
 'V125_09', 'V125_10', 'V125_11', 'V125_12', 'V125_13', 'V125_14', 'V125_15', 'V215_01', 'V215_02', 'V215_03', 'V215_04',
 'V215_05', 'V215_06', 'V215_07', 'V215_08', 'V215_10', 'V215_11', 'V215_12', 'V215_13', 'V215_14', 'V215_15', 'V215_16',
 'V215_17', 'V215_18']

vars_to_drop = []

for variable in high_missings_list:
    if variable not in vars_to_keep:
        
        vars_to_drop.append(variable)

print(vars_to_drop)

['V44_ES', 'V56_NZ', 'V90', 'V91', 'V92', 'V93', 'V94', 'V160A', 'V160B', 'V160C', 'V160D', 'V160E', 'V160F', 'V160G', 'V160H', 'V160I', 'V160J', 'V207A', 'V218_ESMA', 'V217_ESMA', 'V219_ESMA', 'V220_ESMA', 'V221_ESMA', 'V222_ESMA', 'V223_ESMA', 'V224_ESMA', 'MN_35A', 'MN_163A', 'MN_163B', 'MN_163C', 'MN_228L', 'MN_228M', 'MN_228N', 'MN_228O', 'MN_228P', 'MN_228Q', 'MN_228R', 'MN_228S1', 'MN_228S2', 'MN_228S3', 'MN_228S4', 'MN_228S5', 'MN_228S6', 'MN_228S7', 'MN_228S8', 'MN_229A', 'MN_229B', 'MN_230A', 'MN_233A', 'MN_233B', 'MN_234A', 'MN_237A', 'MN_237B1', 'MN_237B2', 'MN_237B3', 'MN_237B4', 'MN_237B5', 'MN_237B6', 'MN_237B7', 'MN_237C1', 'MN_237C2', 'MN_237C3', 'MN_237C4', 'MN_237C5', 'MN_237C6', 'MN_249A1', 'MN_249A2', 'MN_249A3']


In [32]:
len(vars_to_drop)

68

In [ ]:
# drop variables in the vars_to_drop list
cleaning_w6 = 
cleaning_w6.head()

In [ ]:
# quick and dirty stats on some columns
#wvs_w6.describe()